File: java/49.group-anagrams.java
/*
 * @lc app=leetcode id=49 lang=java
 *
 * [49] Group Anagrams
 */

// @lc code=start
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> map = new HashMap();

        for(String s : strs) {
            char[] chars = s.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);

            List<String> list = map.get(key);
            if(list == null) {
                list = new ArrayList();
                map.put(key, list);
            }
            list.add(s);
        }

        return new ArrayList<List<String>>(map.values());
    }
}
// @lc code=end

File: cpp/49.group-anagrams.cpp
/*
 * @lc app=leetcode id=49 lang=cpp
 *
 * [49] Group Anagrams
 */

// @lc code=start
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        std::vector<std::vector<std::string>> res;

        std::unordered_map<std::string, std::vector<std::string>> groups;

        for(std::string& str : strs) {
            std::string key = str;
            std::sort(key.begin(), key.end());

            groups[key].push_back(std::move(str));
        }


        for(auto& kv : groups) {
            res.push_back(std::move(kv.second));
        }

        return res;
    }
};
// @lc code=end

